### PR TITLE
[Web] Base, Util, and Device classes.

### PR DESCRIPTION
**Pull request by jahorton**
Possible TODO:
- [ ] Go through and add some nice access modifiers (`protected`, `private`) where appropriate to the new classes.  Though... `public` is default, and most of those methods should be public...
- [ ] Consider the use of TypeScript namespaces as a protective layer over our new internal classes.  To do this well will likely mean needing to disable Closure's warning about multiple declarations for variables, since said namespaces compile down to objects in JavaScript.

This PR extracts the following into TypeScript classes:  `keyman.util` => `Util` and `keyman.util.device` => `Device`, both contained within the new `kmwutils.ts` source file.  It also extracts `keyman` => `KeymanBase` within `kmwbase.ts`, but only the definition elements original to that file.

These classes are defined in a separate file because of declaration requirements and decorated with type definitions.  TypeScript automatically draws the files into the compilation result **without any significant build script alterations** through use of the `/// <reference path="_____.ts" />` syntax.

Note that without such refactoring, it will be exceedingly difficult to properly specify type data on our core objects in a reliable manner.  The refactoring work will also give us the benefit of enhanced object orientation and the ability to split our source into multiple files as desired.  Oh, and I'm making our usual formatting changes to the affected code while I'm at it.

Speaking of which, related Q:  should `Device` have its own definition file?  It's relatively small and always accessed through the `Util` object, so I went with 'nah' in this implementation.

Significant work has been done to promote the possibility of using Closure's `ADVANCED_OPTIMIZATIONS` flag once completed, ensuring the altered code is as minification-friendly as possible.  Most of the 'magic' toward that end is in the new `kmwapi.ts` file, though important properties must instead be managed in-line with regular code.

### CODE REVIEW

File: web/source/build.sh [hunk 1]
```diff
@@ -170,6 +207,8 @@ if [ $BUILD_EMBED = true ]; then
 
     rm $EMBED_OUTPUT/keyman.js 2>/dev/null
```
Comment (darcywong00) [inline web/source/build.sh]: Maybe add
`echo Compile embedded...` to differentiate 
`echo Typescript compiled` in lines 210 and 234

File: web/samples/uncompiled.html [hunk 2]
```diff
@@ -23,15 +23,7 @@
...
-    <script src="../source/kmwstring.js" type="application/javascript"></script>
-    <script src="../source/kmwbase.js" type="application/javascript"></script>
-    <script src="../source/keymanweb.js" type="application/javascript"></script>
-    <script src="../source/kmwosk.js" type="application/javascript"></script>
-    <script src="../source/kmwnative.js" type="application/javascript"></script>
-    <script src="../source/kmwcallback.js" type="application/javascript"></script>
-    <script src="../source/kmwkeymaps.js" type="application/javascript"></script>
-    <script src="../source/kmwlayout.js" type="application/javascript"></script>
-    <script src="../source/kmwinit.js" type="application/javascript"></script>             
+    <script src="../build/keymanweb.js" type="application/javascript"></script>
```
Comment (darcywong00) [inline web/samples/uncompiled.html]: I thought *.ts gets compiled into  `../build/keymanweb.js`. Maybe the whole page `uncompiled.html` goes away?

Comment (jahorton) [reply web/samples/uncompiled.html]: `../build/keymanweb.js` is the first-stage compiled JavaScript file - it's the direct output of the `tsc` command, before Google Closures performs the second stage.

As a result, no minification has been done upon `../build/keymanweb.js` - the 'fully compiled/minified' version is at `../output/keymanweb.js`.  The latter also has to deal with a second layer of source mapping, which may not be ideal when testing on all browsers.  The lack of minification on the former would prove helpful toward testing for browsers that don't recognize source mapping, so the majority of tests will use this intermediate product.

File: web/source/kmwapi.ts [hunk 3]
```diff
@@ -0,0 +1,44 @@
+/// <reference path="closure.ts" />
+/// <reference path="kmwbase.ts" />
+/// <reference path="kmwutils.ts" />
+
+// Allows proper minification handling.
+/// <reference path="../node_modules/google-closure-library/closure/goog/base.js" />
+
+/**
+ * This file provides calls needed to export our established API elements when performing minification.
+ * Closure can work with these instructions to prevent it from eliminating these names.
```
Comment (jahorton) [inline web/source/kmwapi.ts]: Noting the new issue #518, this file exists to provide those 'extern' declarations and ensure they're preserved during minification.

File: web/source/kmwexthtml.ts [hunk 4]
```diff
@@ -12,6 +13,10 @@ interface Element {
     shim: HTMLElement,                  // Used in subkey elements for smooth fading.
 
     styleSheet: CSSStyleDeclaration,    // Extension of IE.  TS ignores it because of this.
+    attachEvent: any,
```
[the remark this answers is not in the document]
Comment (jahorton) [reply web/source/kmwexthtml.ts]: We can do away with [`attachEvent`](https://developer.mozilla.org/en-US/docs/Web/API/EventTarget/attachEvent) and [`detachEvent`](https://developer.mozilla.org/en-US/docs/Web/API/EventTarget/detachEvent), according to these links and a quick test.  `currentStyle` was even older and is also safe to nix.

I'm having trouble finding anything for `styleSheet`, and while I can find _something_ for `unSelectable`, nothing tells me what versions of IE the legacy property was for.

As for the `Document` properties, we [don't need `selection`](https://caniuse.com/#search=getSelection)!  This causes a bit of a refactoring chain, as there were IE-targeted events (and the handling function) based solely upon document.selection workarounds that are no longer necessary.  It also appears that IE 9+ all support `document.defaultView`, which replaces `document.parentWindow`, meaning even more refactoring comes into play.

**Comment by jahorton** [comment]
Given discussion about how extensive some of this will be, this work will be moved to a separate PR/issue to update to IE9 as a minimum base and investigate the other listed potential deprecation issues: #533.

File: web/source/kmwapi.ts [hunk 5]
```diff
@@ -0,0 +1,44 @@
...
+goog.exportSymbol("Util.prototype.rgba", Util.prototype.rgba);
+goog.exportSymbol("Util.prototype.addStyleSheet", Util.prototype.addStyleSheet);
+goog.exportSymbol("Util.prototype.removeStyleSheet", Util.prototype.removeStyleSheet);
+goog.exportSymbol("Util.prototype.linkStyleSheet", Util.prototype.linkStyleSheet);
+goog.exportSymbol("Util.prototype.addFontFaceStyleSheet", Util.prototype.addFontFaceStyleSheet);
+goog.exportSymbol("Util.prototype.loadCookie", Util.prototype.loadCookie);
+goog.exportSymbol("Util.prototype.saveCookie", Util.prototype.saveCookie);
+goog.exportSymbol("Util.prototype.toNumber", Util.prototype.toNumber);
+goog.exportSymbol("Util.prototype.toFloat", Util.prototype.toFloat);
+goog.exportSymbol("Util.prototype.toNzString", Util.prototype.nzString);
```
[the remark this answers is not in the document]
Comment (jahorton) [reply web/source/kmwapi.ts]: Correct.  The name of the left is the public API name, while the one on the right is the original, private, internal name.  The two need not match, although their base object should.

In the case of `ADVANCED_OPTIMIZATION` with Closure, the one on the right will be minified automatically, with the name restored by the left-hand string.

`goog.exportSymbol` requires use of the Google Closure library, which is a separate package than the Closure compiler.

**Pull request merged**